source/common/tracing/zipkin/zipkin_core_types.cc

```diff
@@ -71,6 +71,14 @@ Annotation& Annotation::operator=(const Annotation& ann) {
   return *this;
 }
 
+void Annotation::changeEndpointServiceName(const std::string& service_name) {
+  if (endpoint_.valid()) {
+    Endpoint endpoint(endpoint_.value());
```

Comment:
That seems to be intuitive, indeed. That's what I did. 
The problem here is different, though. Irrespective of whether or not  `const_cast` is used, the compiler seems to detect an infinite recursion. Perhaps because of optimization? (I am compiling the code with the CI docker image, the one I pulled today).
So, the compiler is not happy about both versions below:
*Version1:*
```
T& value() {
    if (!valid_) {
      throw EnvoyException("fetching invalid Optional value");
    }
    return value_;
}
const T& value() const { return value(); }
```
*Version 2:*
```
T& value() {
    if (!valid_) {
      throw EnvoyException("fetching invalid Optional value");
    }
    return value_;
}
const T& value() const { return const_cast<const T&>(value()); }
```